application.go

```diff
@@ -57,3 +64,120 @@ func (ac *ApplicationClient) GetApplication(
 
 	return res.Application, nil
 }
+
+// GetApplicationsDelegatingToGateway returns the application addresses that are
+// delegating to the given gateway address.
+func (ac *ApplicationClient) GetApplicationsDelegatingToGateway(
+	ctx context.Context,
+	gatewayAddress string,
+	queryHeight uint64,
+) ([]string, error) {
+	allApplications, err := ac.GetAllApplications(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetApplicationsDelegatingToGateway: error getting all applications: %w", err)
+	}
+
+	gatewayDelegatingApplications := make([]string, 0)
+	for _, application := range allApplications {
+		appRing := ApplicationRing{Application: application}
+		// Get the gateways that are delegated to the application
+		// at the query height and check if the given gateway address is in the list.
+		gatewaysDelegatedTo := appRing.ringAddressesAtBlock(queryHeight)
+		if slices.Contains(gatewaysDelegatedTo, gatewayAddress) {
+			// The application is delegating to the given gateway address, add it to the list.
+			gatewayDelegatingApplications = append(gatewayDelegatingApplications, application.Address)
+		}
+	}
+
+	return gatewayDelegatingApplications, nil
+}
+
+type ApplicationRing struct {
+	types.Application
+	PublicKeyFetcher
+}
+
+// GetRing returns the ring for the application.
+// The ring is created using the application's public key and the public keys of
+// the gateways that are currently delegated from the application.
+func (a ApplicationRing) GetRing(
```

Comment:
I think `GetRing` should use the logic provided by `poktroll`. I understand that for now it implies getting the `poktroll` `RingClient` that comes with some unwanted "client" dependencies.

I can have a PR making `poktroll`'s `GetRingForAddressAtHeight` a pure function. Meanwhile, we can keep this.

Reply:
Great idea: `poktroll` is a better fit for this function. Please see my suggestion on potentially enhancing the `Session` struct to build the application's ring.